sympy/functions/special/bessel.py

```diff
@@ -610,43 +629,62 @@ class jn(SphericalBesselBase):
 
     where :math:`J_\nu(z)` is the Bessel function of the first kind.
 
+    The spherical Bessel functions of integral order are
+    calculated using the formula:
+
+    .. math:: j_n(z) = f_n(z) \sin{z} + (-1)^{n+1} f_{-n-1}(z) \cos{z},
+
+    where the coefficients :math:`f_n(z)` are available as
+    :func:`polys.orthopolys.spherical_bessel_fn`.
+
     Examples
     ========
 
-    >>> from sympy import Symbol, jn, sin, cos, expand_func
+    >>> from sympy import Symbol, jn, sin, cos, expand_func, besselj, bessely
+    >>> from sympy import simplify
     >>> z = Symbol("z")
-    >>> print(jn(0, z).expand(func=True))
+    >>> nu = Symbol("nu", integer=True)
+    >>> print(expand_func(jn(0, z)))
     sin(z)/z
-    >>> jn(1, z).expand(func=True) == sin(z)/z**2 - cos(z)/z
+    >>> expand_func(jn(1, z)) == sin(z)/z**2 - cos(z)/z
     True
     >>> expand_func(jn(3, z))
     (-6/z**2 + 15/z**4)*sin(z) + (1/z - 15/z**3)*cos(z)
-
-    The spherical Bessel functions of integral order
-    are calculated using the formula:
-
-    .. math:: j_n(z) = f_n(z) \sin{z} + (-1)^{n+1} f_{-n-1}(z) \cos{z},
-
-    where the coefficients :math:`f_n(z)` are available as
-    :func:`polys.orthopolys.spherical_bessel_fn`.
+    >>> jn(nu, z).rewrite(besselj)
+    sqrt(2)*sqrt(pi)*sqrt(1/z)*besselj(nu + 1/2, z)/2
+    >>> jn(nu, z).rewrite(bessely)
+    (-1)**nu*sqrt(2)*sqrt(pi)*sqrt(1/z)*bessely(-nu - 1/2, z)/2
+    >>> jn(2, 5.2+0.3j).evalf(20)
+    0.099419756723640344491 - 0.054525080242173562897*I
 
     See Also
     ========
 
     besselj, bessely, besselk, yn
 
+    References
+    ==========
+
+    .. [1] http://dlmf.nist.gov/10.47
+
     """
 
     def _rewrite(self):
         return self._eval_rewrite_as_besselj(self.order, self.argument)
 
+    @assume_integer_order
     def _eval_rewrite_as_besselj(self, nu, z):
-        return sqrt(pi/(2*z)) * besselj(nu + S('1/2'), z)
+        return sqrt(pi/(2*z)) * besselj(nu + S.Half, z)
+
+    @assume_integer_order
+    def _eval_rewrite_as_bessely(self, nu, z):
+        return (-1)**nu * sqrt(pi/(2*z)) * bessely(-nu - S.Half, z)
+
+    def _eval_rewrite_as_yn(self, nu, z):
```

Comment:
I am inclined to prefer 1) provided it can be reasonably implemented. Perhaps it would not be too big a restriction to always demand a positive response to ``nu.is_integer``. It would be true for explicit integers and, if symbols are desired, for ``Symbol('n', integer=True)``.

Then there remains the question what to do if the order is not an integer. In most cases SymPy is very conservative about preserving information. So functions (and other expressions) that are actually undefined are often returned by the ctor as unevaluated expressions. (That is the reason why it was possible to formally "rewrite" ``jn(S.Half, x)`` even if it is meaningless.)

Instead, I think we should have a ``class Undefined`` that would have the (meaningless) expression as an attribute and would print e.g. as ``Undefined('jn(S.Half, x)')``.

That should be returned by the ctor.